### PR TITLE
Switch to using nulldb for tasks that need a database

### PR DESCRIPTION
**Pull request by ehelms**
@tbrisker @mmoll @ekohl @evgeni 

I had a successful Foreman + foreman_ansible build with using the nulldb adapter. This shows it working for RPM build process and can be cleaned up. I wanted y'all to see it is possible and the changes I had to make.

The assumptions we make today are that `database.yml.example` has a basic sqlite database defined by it. If we drop Sqlite (which I think we should to avoid having to keep SQL compatiable with it in our applications and drop more out of our test matrix) we will need to decide what the example and default database.yml deployed is (perhaps "basic" postgresql configuration is the right answer).

Note that nulldb requires a schema.rb file to exist, but the contents of that file do not matter.

I would suggest that we also modify CI to run asset precompile and apipie cache testing using nulldb to ensure that it continues to work rather than catching it down in the RPM layer if a problem arises.

This has a nice advantage as well of not needing to do any database migrating.

To summarize:

 * nulldb approach works
 * nulldb feels clean and reduces some of the RPM build overhead
 * we should test this in CI IMO and should decide if we want to include it in foreman core codebase or just how I have it here for build tasks

### CODE REVIEW

File: package_manifest.yaml [hunk 1]
```diff
@@ -59,12 +59,14 @@ foreman_scl_packages:
         - "jenkins_job=foreman-develop-source-release"
         - "jenkins_url=https://ci.theforeman.org"
       build_package_tito_builder: "tito.builder.FetchBuilder"
+      nightly_githash: '1'
```
Comment (ekohl) [inline package_manifest.yaml:62]: Debug

File: mock/el7-scl.cfg [hunk 2]
```diff
@@ -6,6 +6,7 @@ config_opts['dist'] = 'el7'  # only useful for --resultdir variable subst
 # beware RHEL use 7Server or 7Client
 config_opts['releasever'] = '7'
 config_opts['package_manager'] = 'yum'
+config_opts['yum_install_command'] += ' --disablerepo sclo*'
```
Comment (ekohl) [inline mock/el7-scl.cfg:9]: Mind separating this to a separate commit? https://github.com/theforeman/foreman-packaging/pull/5047 was intended to solve the same issue

File: packages/foreman/foreman/foreman.spec [hunk 3]
```diff
@@ -245,14 +245,16 @@ BuildRequires: %{?scl_prefix}rubygem(graphql) < 1.9.0
...
+
 # assets
 %if 0%{?scl:1}
 BuildRequires: %{scl}-runtime-assets >= 6
 BuildRequires: %{scl}-runtime-assets < 7
 %else
 BuildRequires: nodejs-packaging
 %endif
-BuildRequires: %{?scl_prefix_nodejs}nodejs >= 6.10
+BuildRequires: %{?scl_prefix_nodejs}nodejs
```
Comment (ekohl) [inline packages/foreman/foreman/foreman.spec:257]: I actually would like to set this to the correct version (is that 10 now?)

File: packages/foreman/foreman/foreman.spec [hunk 4]
```diff
@@ -725,7 +728,17 @@ make -C locale all-mo
...
+cat <<DATABASE > config/database.yml
+production:
+  adapter: nulldb
+development:
+  adapter: nulldb
+DATABASE
+cat <<DB_SCHEMA > db/schema.rb
+ActiveRecord::Schema.define(version: 1) do
+end
+DB_SCHEMA
```
Comment (ekohl) [inline packages/foreman/foreman/foreman.spec:741]: As you suggested, I think it'd be good if this lived in Foreman itself as `bundler.d/nulldb.rb` and `config/database.yml.nulldb`. Not sure about the schema.

File: packages/foreman/foreman/foreman.spec [hunk 5]
```diff
@@ -792,7 +804,7 @@ find %{buildroot}%{_datadir}/%{name} -type d -name "test" |xargs rm -rf
 rm -f %{buildroot}%{_datadir}/%{name}/bin/spring
 
 # Move config files to %{_sysconfdir}
-mv %{buildroot}%{_datadir}/%{name}/config/database.yml.example %{buildroot}%{_datadir}/%{name}/config/database.yml
+#mv %{buildroot}%{_datadir}/%{name}/config/database.yml %{buildroot}%{_datadir}/%{name}/config/database.yml
```
Comment (ekohl) [inline packages/foreman/foreman/foreman.spec:807]: This is IMHO still good for fresh installs to keep

File: packages/foreman/foreman/foreman.spec [hunk 6]
```diff
@@ -885,8 +897,11 @@ cp -r %%{%{name}_dir} ./%{_datadir} || echo 0 \\
 mkdir -p ./%{_localstatedir}/lib/%{name} \\
 cp -r %{_localstatedir}/lib/%{name}/db ./%{_localstatedir}/lib/%{name} || echo 0 \\
 unlink ./%{_datadir}/%{name}/db \\
-ln -sv \`pwd\`/%{_localstatedir}/lib/%{name}/db ./%{_datadir}/%{name}/db \\
 pushd ./%%{%{name}_dir} \\
+mkdir db/ \\
+cp -rf %{_datadir}/%{name}/db/* db/ \\
```
Comment (ekohl) [inline packages/foreman/foreman/foreman.spec:902]: Doesn't this work:
```suggestion
cp -rf %{_datadir}/%{name}/db ./ \\
```

File: mock/el7-scl.cfg (hunk 2 above)
Comment (ehelms) [reply mock/el7-scl.cfg:9]: This is not the final commit. I just kept things in here I need present to allow me to test and re-test.

File: packages/foreman/foreman/foreman.spec (hunk 5 above)
Comment (ehelms) [reply packages/foreman/foreman/foreman.spec:807]: This is purely to allow testing this.

File: packages/foreman/foreman/foreman.spec (hunk 6 above)
Comment (ehelms) [reply packages/foreman/foreman/foreman.spec:902]: I don't know. But this may change depending on how much of this nulldb code we put into Foreman. I wouldn't focus too much on the code itself as much as the "process" that is required. I'll open a Foreman PR with some of the proposed idea changes to help.

**Comment by evgeni** [comment]
Oh, one thing, the nulldb readme does not mention AR 6.0, do you know if it still works with Rails6?

**Comment by tbrisker** [comment]
I like this! 

> Oh, one thing, the nulldb readme does not mention AR 6.0, do you know if it still works with Rails6?

Looks like they forgot to update the readme, but version 0.4.0 does seem to include AR 6 support: https://github.com/nulldb/nulldb/tree/v0.4.0

**Comment by ohadlevy** [comment]
nice - would you mind fixing the dockerfile too? :)

**Comment by ehelms** [comment]
This currently includes a fix related to https://github.com/theforeman/foreman-packaging/pull/5078 but I'd prefer that #5078 go in first and then I update this one.

File: packages/foreman/foreman/foreman.spec [hunk 7]
```diff
@@ -252,7 +253,7 @@ BuildRequires: %{scl}-runtime-assets < 7
 %else
 BuildRequires: nodejs-packaging
 %endif
-BuildRequires: %{?scl_prefix_nodejs}nodejs >= 6.10
+BuildRequires: %{?scl_prefix_nodejs}nodejs >= 10
```
Comment (evgeni) [inline packages/foreman/foreman/foreman.spec:256]: doesn't feel like it belongs here but :woman_shrugging:

File: packages/foreman/foreman/foreman.spec [hunk 8]
```diff
@@ -885,22 +883,23 @@ cp -r %%{%{name}_dir} ./%{_datadir} || echo 0 \\
...
+mkdir db/ \\
+cp -rf %{_datadir}/%{name}/db/* db/ \\
+mv db/schema.rb.nulldb db/schema.rb
 \\
 ln -s %{?scl_prefix_nodejs:%{_scl_root}}%{nodejs_sitelib} node_modules \\
 export GEM_PATH=%%{buildroot}%%{gem_dir}:\${GEM_PATH:+\${GEM_PATH}}\${GEM_PATH:-\`%{?scl:scl enable %%{scl} -- }ruby -e "print Gem.path.join(':')"\`} \\
 unlink tmp \\
 \\
 rm \`pwd\`/config/initializers/encryption_key.rb \\
+rm \`pwd\`/config/database.yml \\
```
Comment (evgeni) [inline packages/foreman/foreman/foreman.spec:896]: what would have created this file?

Comment (ehelms) [reply packages/foreman/foreman/foreman.spec:896]: This line copies the example to `database.yml` so there is a default: https://github.com/theforeman/foreman-packaging/pull/5058/files#diff-3399e5c03dc282964481f8f3ced9c141R796

When we drop SQlite we need to re-think this but for now we are just switching the nulldb usage.

**Comment by evgeni** [comment]
@ehelms #5078 is **IN**

File: packages/foreman/foreman/foreman.spec (hunk 8 above)
Comment (evgeni) [reply packages/foreman/foreman/foreman.spec:896]: that link goes nowhere, stupid GH?

**Comment by tbrisker** [comment]
We need source-release to finish, 
`cp: cannot stat 'db/schema.rb.nulldb': No such file or directory`

**Comment by mmoll** [comment]
[test rpm]
